text/2017-07-23-relaxed-memory.md

```diff
@@ -0,0 +1,218 @@
+# Summary
+
+This is a proposal for clarifying the synchronization performed in the epoch-based memory
+reclamation scheme, and explaining why Crossbeam's implementation is correct
+w.r.t. C11/C++11/LLVM/Rust relaxed-memory concurrency model (the "C11 memory model" from now on).
+
+
+
+# Motivation
+
+Crossbeam is based on the epoch-based memory reclamation scheme (EBR).  In EBR, an **epoch** (think:
+timestamp) is maintained, and a location is mark with the current epoch when the location is
+unlinked.  The unlinked location is safe to deallocate later when the epoch is advanced enough.  As
+you may guess, the correctness of EBR crucially depends on the invariant on the epoch: it is
+advanced only when old locations unlinked in the past are no longer accessible.
+
+EBR is introduced in [Keir Fraser's PhD thesis][kfraser], and later Aaron Turon introduced it
+in [his blog post on Crossbeam][aturon] to the Rust community.  Both explains EBR in details so that
+a reader can understand not only the intuition behind it but also why it is correct.  However, both
+assumed we are on the *sequentially consistent (SC) concurrency model*: every access to the memory
+from all the threads are *linearizable*, and each load reads from the most-recently stored value
+according to the linearization order.  However, unfortunately, the real-world doesn't allow us to
+enjoy the luxury of SC.  In particular, the C11/C++11/LLVM/Rust memory model is *relaxed*, or
+*weak*, allowing each load to read from stale old values.
+
+As a result, it is still unclear after two years of development whether the implementation of
+Crossbeam is correct w.r.t. the C11 memory model.  We believe that in order for Crossbeam to indeed
+serve as the crossbeam for Rust concurrency, as `java.util.concurrent` does for Java concurrency,
+Crossbeam's correctness should be formally analyzed and hopefully verified.  This RFC aims for
+providing a step towards the goal.
+
+## Non-motivation
+
+Crossbeam is the marriage of EBR and Rust's linear type system with lifetime.  However, this RFC
+does not aim for clarifying how Rust's type system helps to ensure the correctness of Crossbeam.
+
+
+
+# Detailed design
+
+In this section, we present the proposed implementation of Crossbeam, and explain why it is correct
```

Comment:
I would rather not call it a proof, as there are many gaps..  For concurrency, without actually formallying  proving it, I am hesitant to say it is a proof..